_posts/2021-09-13-user-defined-value-types.md

```diff
@@ -0,0 +1,152 @@
+`Price.unwrap` and `Quantity.unwrap` respectively.
+
+The data-representation of the values of such types are inherited from the underlying type and the
+underlying type is also used in the ABI. This means that the following two `transfer` functions
+would be identical, i.e., they have the same [function
+selector](https://docs.soliditylang.org/en/latest/abi-spec.html#function-selector) as well as the
+same [ABI encoding and
+decoding](https://docs.soliditylang.org/en/latest/abi-spec.html#mapping-solidity-to-abi-types). This
+allows using user defined value types in a backwards compatible way.
+
+```solidity
+pragma solidity ^0.8.8;
+
+type Decimal18 is uint256;
+
+interface MinimalERC20 {
+    function transfer(address to, Decimal18 value) external;
+}
+
+interface AnotherMinimalERC20 {
+    function transfer(address to, uint256 value) external;
+}
+```
+
+Notice in the above example, how the user defined type `Decimal18` makes it clear that a value is
+supposed to represent a number with 18 decimals.
+
+## Example
+
+The following example illustrates a custom type `UFixed256x18` representing a decimal fixed point
+type with 18 decimals and a minimal library to do arithmetic operations on the type.
+
+```solidity
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity ^0.8.8;
+
+// Represent a 18 decimal, 256 bit wide fixed point type using a user defined value type.
+type UFixed256x18 is uint256;
+
+/// A minimal library to do fixed point operations on UFixed256x18.
+library FixedMath {
+    uint constant multiplier = 10**18;
+
+    /// Adds two UFixed256x18 numbers. Reverts on overflow, relying on checked
+    /// arithmetic on uint256.
+    function add(UFixed256x18 a, UFixed256x18 b) internal pure returns (UFixed256x18) {
+        return UFixed256x18.wrap(UFixed256x18.unwrap(a) + UFixed256x18.unwrap(b));
+    }
+    /// Multiplies UFixed256x18 and uint256. Reverts on overflow, relying on checked
+    /// arithmetic on uint256.
+    function mul(UFixed256x18 a, uint256 b) internal pure returns (UFixed256x18) {
+        return UFixed256x18.wrap(UFixed256x18.unwrap(a) * b);
+    }
+    /// Take the floor of a UFixed256x18 number.
+    /// @return the largest integer that does not exceed `a`.
+    function floor(UFixed256x18 a) internal pure returns (uint256) {
+        return UFixed256x18.unwrap(a) / multiplier;
+    }
+    /// Turns a uint256 into a UFixed256x18 of the same value.
+    /// Reverts if the integer is too large.
+    function toUFixed256x18(uint256 a) internal pure returns (UFixed256x18) {
+        return UFixed256x18.wrap(a * multiplier);
+    }
+}
+```
+
+Notice how `UFixed256x18.wrap` and `FixedMath.toUFixed256x18` have the same signature but perform
+two very different operations: The `UFixed256x18.wrap` function returns a `UFixed256x18` that has
+the same data representation as the input, whereas `toUFixed256x18` returns a `UFixed256x18` that
+has the same numerical value.
+
```

Comment:
Maybe we could also add that a proper type implementation should only used the `wrap` and `unwrap` functions in the file that defines the type but not outside to provide encapsulation?

Reply:
I can add this, but this makes me wonder if this should have been part of the compiler, i.e., `wrap` and `unwrap` functions are only available in the same scope as the definition.